### PR TITLE
WIP: Retrofit converter and sealed class for response types

### PR DESCRIPTION
**Pull request by akwc10**

### CODE REVIEW

File: app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt [hunk 1]
```diff
@@ -13,16 +13,21 @@ interface JourneyPlannerApiService {
...
 
     companion object {
         private const val TFL_API_BASE_URL = "https://api.tfl.gov.uk/"
-        private val client = OkHttpClient().newBuilder().addInterceptor(ParametersInterceptor()).build()
+        private val client = OkHttpClient().newBuilder()
+            .addInterceptor(ParametersInterceptor())
+            .build()
 
-        fun createApiService(): JourneyPlannerApiService = Retrofit.Builder()
+        val retrofit: Retrofit = Retrofit.Builder()
```
Comment (fsbarata) [inline app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt]: is this change necessary? what have we improved here? Specially concerned about making it public

File: app/src/main/java/com/my/journeyplanner/helpers/ResponseCodeInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.my.journeyplanner.helpers
+
+import okhttp3.Interceptor
+import okhttp3.Response
+
+class ResponseCodeInterceptor : Interceptor {
```
Comment (fsbarata) [inline app/src/main/java/com/my/journeyplanner/helpers/ResponseCodeInterceptor.kt]: Better call it something we know what it is. Say like, MultiOptionResponseCodeInterceptor or something

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt [hunk 3]
```diff
@@ -20,30 +24,91 @@ class MainPresenter(private val view: MainContract.View) : MainContract.Presente
...
+                throwable: Throwable
+            ) {
+                if (throwable is IOException) {
+                    Log.d(TAG, "*** Network failure ***")
+                } else {
+                    Log.d(TAG, "*** Conversion issue ***")
+                }
             }
 
             override fun onResponse(
```
Comment (fsbarata) [inline app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: this method is wayyyy too big. Start breaking it down

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt [hunk 4]
```diff
@@ -20,30 +24,91 @@ class MainPresenter(private val view: MainContract.View) : MainContract.Presente
...
             ) {
-                Log.i(TAG, "response: $response")
-                response.errorBody()?.apply {
-                    Log.i(TAG, "errorBody().source(): ${source()}")
-                    Log.i(TAG, "errorBody().charStream().readText(): ${charStream().readText()}")
+                if (response.isSuccessful) {
+                    Log.d(TAG, "*** ItineraryResult: ${response.body()} ***")
+//                TODO("Display data in new activity")
+                } else when (response.code()) {
+                    300 -> {
```
Comment (fsbarata) [inline app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: didn't we make sure this doesn't happen?

File: app/src/main/java/com/my/journeyplanner/helpers/StringUtils.kt [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+fun String.countOccurrences(searchedString: String, index: Int = 0): Int {
+    if (index >= this.length) return 0
+
+    val thisStringIndex = this.indexOf(searchedString, index)
+    val count = if (thisStringIndex > 0) 1 else 0
+    val thisStringIndexIncremented =
+        if (count > 0) index + thisStringIndex + searchedString.length else this.length
+
+    return count + countOccurrences(searchedString, thisStringIndexIncremented)
```
Comment (fsbarata) [inline app/src/main/java/com/my/journeyplanner/helpers/StringUtils.kt]: or you can scrap all this code and simply:

fun String.countOccurrences(searchedString: String) = split(searchString).length - 1

Also write unit tests before you refactor.

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt [hunk 6]
```diff
@@ -20,30 +24,91 @@ class MainPresenter(private val view: MainContract.View) : MainContract.Presente
...
         )
 
-/*            TODO("Default adapter treats response code 300 Multiple Options as unsuccessful and")
-            TODO("puts response in the errorBody but is actually a valid response.")
-            TODO("Implement custom adapter to parse the valid response in the errorBody")   */
-        call!!.enqueue(object : Callback<JourneyPlannerDisambiguationResult> {
-            override fun onFailure(call: Call<JourneyPlannerDisambiguationResult>, throwable: Throwable) {
-                Log.e(TAG, "Other error: ${throwable.message}")
-                Log.e(TAG, "Other error: ${throwable.stackTrace.forEach { println(it) }}")
+        call!!.enqueue(object : Callback<JourneyPlanner.ItineraryResult> {
```
Comment (fsbarata) [inline app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: presenters are about usecase->view interaction.

You're mixing the usecase layer here. Better is to ask for data to the usecase, and the use case deals with the api and all its particularities. Presenters just care about wiring logic

File: app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt (hunk 1 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt]: Used to access JourneyPlannerApiService.retrofit.responseBodyConverter().
I can use fakeRetrofit:
        val fakeRetrofit: Retrofit = Retrofit.Builder()
            .baseUrl("https://ttttt.com/")
            .client(OkHttpClient().newBuilder().build())
            .addConverterFactory(MoshiConverterFactory.create())
            .build()
Am I missing the point?

File: app/src/main/java/com/my/journeyplanner/helpers/ResponseCodeInterceptor.kt (hunk 2 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/helpers/ResponseCodeInterceptor.kt]: Updated as advised

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt (hunk 4 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: Yes, but I removed the MultiOptionResponseCodeInterceptor() as I couldn't get it to work with:
public @Nullable Converter<ResponseBody, ?> responseBodyConverter(Type type,
        Annotation[] annotations, Retrofit retrofit)
due to the generic return type ResponseBody i.e. I can't do this:
                                val converter: Converter<T, JourneyPlanner.NotIdentifiedResult> =
                                    JourneyPlannerApiService.fakeRetrofit.responseBodyConverter(
                                        JourneyPlanner.NotIdentifiedResult::class.java,
                                        arrayOfNulls<Annotation>(0)
                                    )
  converter.convert(response.body())
where:
  public @Nullable T body() {}
There'll be an obvious way but I haven't found it

File: app/src/main/java/com/my/journeyplanner/helpers/StringUtils.kt (hunk 5 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/helpers/StringUtils.kt]: Feedback implemented

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt (hunk 3 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: De-duped and functionalised

File: app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt (hunk 1 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt]: Renamed to MockRetrofit using MockInterceptor() for responseBodyConverter(). I'd moved this into a separate class but I suspect you have a better solution!

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt (hunk 6 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: Added uses cases but I'm not convinced with my implementation

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt (hunk 4 above)
Comment (fsbarata) [reply app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: yeah I'll help you out on this later

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt (hunk 6 above)
Comment (fsbarata) [reply app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: No, you need something of the sort of a Repository / Provider / etc. I'll give you some ideas of what to do later

Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: Added repository, datasource and use cases/interactors.

File: app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt (hunk 1 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/helpers/JourneyPlannerApiService.kt]: Removed as no longer required

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt (hunk 3 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: Logic no longer required due to implementation of MultiOptionResponseCodeInterceptor() and JourneyPlannerResultConverterFactory

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt (hunk 4 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: Reinstated MultiOptionResponseCodeInterceptor()

File: app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt (hunk 6 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/presenters/MainPresenter.kt]: Fixed repository which was just delegating to datasource and removed latter

File: app/src/main/java/com/my/journeyplanner/framework/JourneyPlannerResultConverterFactory.kt [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+                jsonBody.hasLocationDisambiguationKey(
+                    hasToLocationDisambiguation,
+                    TO_LOCATION_DISAMBIGUATION,
+                    DISAMBIGUATION_OPTIONS
+                )
+            val isFromAndToLocationDisambiguationOptions =
+                hasFromLocationDisambiguationOptions && hasToLocationDisambiguationOptions
+            val isFromDisambiguationOptions =
+                hasFromLocationDisambiguationOptions && !hasToLocationDisambiguationOptions
+            val isToDisambiguationOptions =
```
Comment (fsbarata) [inline app/src/main/java/com/my/journeyplanner/framework/JourneyPlannerResultConverterFactory.kt]: from what I gathered, there was a field called type which had a full qualified type. What happened to that and why did you decide to understand what the data was comprised of?

File: app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    key: String
+): Boolean {
+    try {
+        if (hasLocationDisambiguation) {
+            if (this.getJSONObject(locationDisambiguation).has(key)) {
+                when (key) {
+                    DISAMBIGUATION_OPTIONS -> {
+                        if (this.getJSONObject(locationDisambiguation).getJSONArray(
+                                key
+                            ).length() > 0
```
Comment (fsbarata) [inline app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt]: isNotEmpty()

File: app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+                            ) == NOT_IDENTIFIED
+                        ) return true
+                    }
+                }
+            } else {
+                Log.d(TAG, "Missing key: $locationDisambiguation.$key")
+            }
+        }
+    } catch (e: JSONException) {
+        Log.e(TAG, e.message)
```
Comment (fsbarata) [inline app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt]: not cool. Why can it go wrong?

File: app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                    }
+                    MATCH_STATUS -> {
+                        if (this.getJSONObject(locationDisambiguation).getString(
+                                key
+                            ) == NOT_IDENTIFIED
+                        ) return true
+                    }
+                }
+            } else {
+                Log.d(TAG, "Missing key: $locationDisambiguation.$key")
```
Comment (fsbarata) [inline app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt]: what's the problem with this?

File: app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt (hunk 8 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt]: JSONArray doesn't implement Collection hence can't apply this

File: app/src/main/java/com/my/journeyplanner/framework/JourneyPlannerResultConverterFactory.kt (hunk 7 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/framework/JourneyPlannerResultConverterFactory.kt]: ItineraryResult i.e. exact match
{"$type":"Tfl.Api.Presentation.Entities.JourneyPlanner.ItineraryResult, Tfl.Api.Presentation.Entities",...
notidentified i.e. no match either way
{"$type":"Tfl.Api.Presentation.Entities.JourneyPlanner.DisambiguationResult, Tfl.Api.Presentation.Entities",...
From To Disambiguation i.e. multiple options both ways
{"$type":"Tfl.Api.Presentation.Entities.JourneyPlanner.DisambiguationResult, Tfl.Api.Presentation.Entities",...
From DisambiguationResult i.e. mutliple options in from direction
{"$type":"Tfl.Api.Presentation.Entities.JourneyPlanner.DisambiguationResult, Tfl.Api.Presentation.Entities",...
To DisambiguationResult i.e. multiple options in to direction
{"$type":"Tfl.Api.Presentation.Entities.JourneyPlanner.DisambiguationResult, Tfl.Api.Presentation.Entities",...

The type is the same for the other types with the exception of ItineraryResult hence why I have to drill down a bit more

File: app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt (hunk 9 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt]: It shouldn't throw an exception as I've already checked the keys exists beforehand. Just being overcautious. Removed

File: app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt (hunk 10 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/framework/JsonOperations.kt]: Just in case TFL decide to change the API. Remove it?

File: app/src/main/java/com/my/journeyplanner/framework/JourneyPlannerResultConverterFactory.kt (hunk 7 above)
Comment (akwc10) [reply app/src/main/java/com/my/journeyplanner/framework/JourneyPlannerResultConverterFactory.kt]: Reduced data types from 5 to 2 and removed associated logic
